File: python/알고리즘/pro/lev2_짝지어제거하기.py
# 효율성 0점. 배열 중간에 pop하는 과정이 O(N)이라서 그런듯함.
# 스택으로 하면?
def solution(s):
    answer = 1
    a = list(s)
    i = 0
    while i < len(a)-1:
        if len(a) == 0:
            break
        if a[i] == a[i+1]:
            a.pop(i)
            a.pop(i)
            if i > 0:
                i -= 1
        else:
             i += 1
    if len(a) != 0:
        answer = 0
    return answer

# 좋다. 효율성 100점!
def solution(s):
    answer = 1
    stack = []
    i = 0
    while i < len(s):
        if stack:
            if stack[-1] == s[i]:
                stack.pop()
            else:
                stack.append(s[i])
        else:
            stack.append(s[i])
        i += 1
    if stack:
        answer = 0
    
    return answer

File: python/알고리즘/pro/lev2_짝지어제거하기.swift
import Foundation

func solution(_ s:String) -> Int{
    var answer:Int = -1

    // [실행] 버튼을 누르면 출력 값을 볼 수 있습니다.
    var i = 0
    var sArray = Array(s)
    var stack = [Character]()
    while i < sArray.count {
        if stack.count != 0 {
            // last로 마지막 놈을 가져오거나
            // endIndex - 1로 인덱싱 하거나
            // count - 1로 인덱싱 하거나
            if stack.last! == sArray[i] {
                stack.removeLast()
            } else {
                stack.append(sArray[i])
            }
        } else {
            stack.append(sArray[i])
        }
        i += 1
    }
    if stack.count == 0 {
        answer = 1
    } else {
        answer = 0
    }

    return answer
}